standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -624,6 +624,40 @@ public boolean openTransaction(String isolationLevel) {
     return result;
   }
 
+  @Override
+  public int updateParameterWithExpectedValue(Table table, String key, String expectedValue, String newValue)
+      throws MetaException {
+    String dml = String.format(
```

Comment:
Could we do something like this:
https://www.datanucleus.org/products/accessplatform_6_0/jdo/query.html#jdoql_bulkupdate

I would like to avoid:
- Writing strings to the queries based on table parameters
- Using native connection

If using jdo queries doesn't work, then I would create a method in `MetaStoreDirectSql` for this, and throw an exception if directSql is not turned on

Reply:
Setting `datanucleus.query.jdoql.allowAll=true` allows us to run UPDATE with JDO query. I tried something like this but it doesn't work.
```java
    Map<String, String> newParams = new HashMap<>(table.getParameters());
    newParams.put(key, newValue);
    openTransaction();
    Query query = pm.newQuery("UPDATE org.apache.hadoop.hive.metastore.model.MTable " +
        "SET parameters=:newparams WHERE database.name == :dbname && tableName == :tblname && " +
        "parameters.containsEntry(:key, :expval)");
    int affectedRows = (int) query.executeWithMap(ImmutableMap.of(
        "newparams", newParams,
        "dbname", table.getDbName(),
        "tblname", table.getTableName(),
        "key", key,
        "expval", expectedValue
    ));
```
The error is
```
Caused by: org.datanucleus.store.rdbms.sql.expression.IllegalExpressionOperationException: Cannot perform operation "==" on org.datanucleus.store.rdbms.sql.expression.MapExpression@6f9c5048 and org.datanucleus.store.rdbms.sql.expression.MapLiteral@5114b7c7
	at org.datanucleus.store.rdbms.sql.expression.SQLExpression.eq(SQLExpression.java:381)
	at org.datanucleus.store.rdbms.sql.expression.MapExpression.eq(MapExpression.java:80)
	at org.datanucleus.store.rdbms.query.QueryToSQLMapper.compileUpdate(QueryToSQLMapper.java:1134)
```
I believe it's because MapExpression only supports [eq with null literals in DN](https://github.com/datanucleus/datanucleus-rdbms/blob/master/src/main/java/org/datanucleus/store/rdbms/sql/expression/MapExpression.java#L71).

Reply:
is that ok, if i change code as below?
`   String queryText =  "UPDATE \"TABLE_PARAMS\" SET \"PARAM_VALUE\" = ? " +
            "WHERE \"TBL_ID\" in (select TBL_ID from TBLS join DBS ON DBS.DB_ID=TBLS.DB_ID WHERE NAME=? and TBL_NAME=? ) AND \"PARAM_KEY\" = ? AND \"PARAM_VALUE\" = ?";
    try (QueryWrapper query = new QueryWrapper(pm.newQuery("javax.jdo.query.SQL", queryText))) {
       Object[] params = new Object[4];
       params[0] = ...;
       long res = query.executeWithArray(params);
...
    }`
cause i don't cherry pick [HIVE-22234](https://issues.apache.org/jira/browse/HIVE-22234), can't get table id

Reply:
Yes, for old versions where Table object doesn't have the ID, I think we need to retrieve it from `TBLS`. I'm planning to do this in PRs for the release branches.